### PR TITLE
feat: add flexible handshake configuration

### PR DESCRIPTION
**Pull request by Rqnsom**
- handshake procedure slightly refactored
- netprio test aligned with the new solution

### CODE REVIEW

File: src/protocol/handshake.rs [hunk 1]
```diff
@@ -117,50 +167,49 @@ impl Handshake for InnerNode {
...
+                rsp_header(format!("X-Algorand-Version: {}", cfg.ar_accept_version));
+                rsp_header(format!("X-Algorand-Genesis: {}", cfg.ar_genesis));
+                if let Some(ref challenge) = cfg.challenge {
+                    rsp_header(format!("X-Algorand-Prioritychallenge: {challenge}"));
+                }
+                rsp_header("".into()); // A HTTP header ends with '\r\n'
 
-                info!(parent: self.node().span(), "sending handshake message: {:?}", rsp);
+                let rsp = Bytes::from(rsp);
+                info!(parent: self.node().span(), "sending handshake a handshake response: {:?}", rsp);
```
Comment (asmie) [inline src/protocol/handshake.rs:212]: Shouln't be "sending a handshake response" ?

**Review by asmie — COMMENTED** [review]
LGTM but need to clean clippy suggestions.

File: src/protocol/handshake.rs [hunk 2]
```diff
@@ -28,9 +28,10 @@ const X_AG_ALGORAND_GENESIS: &str = "private-v1";
 // 0x0a 0x0b 0x0c 0x0d 0x0e 0x0f 0x10, the value of the header
 // field would be "AQIDBAUGBwgJCgsMDQ4PEC=="
 /// A key-accept pair for a Sec-WebSocket-Key header.
-struct SecWebSocket {
-    key: String,
-    accept: String,
+#[derive(Clone, Debug)]
+pub struct SecWebSocket {
+    pub key: String,
```
[the remark this answers is not in the document]
Comment (Rqnsom) [reply src/protocol/handshake.rs:34]: Done

File: src/protocol/handshake.rs (hunk 1 above)
Comment (Rqnsom) [reply src/protocol/handshake.rs]: Done

File: src/protocol/handshake.rs [hunk 3]
```diff
@@ -42,55 +43,104 @@ impl SecWebSocket {
...
-                req.extend_from_slice(
-                    format!("X-Algorand-Version: {}\r\n", X_AG_ALGORAND_VERSION).as_bytes(),
-                );
-                req.extend_from_slice(
-                    format!("X-Algorand-Genesis: {}\r\n", X_AG_ALGORAND_GENESIS).as_bytes(),
-                );
-                req.extend_from_slice(b"\r\n");
-                let req = Bytes::from(req);
+                let mut req_header = |mut header: String| {
+                    header.push_str("\r\n");
```
[the remark this answers is not in the document]
Comment (asmie) [reply src/protocol/handshake.rs:113]: I thought it was adding \r\n after the header (push_str append given str to the value) - so it should work as expected - there is a header and we append here just \r\n.
But I was thinking simillar yesterday :)

File: src/protocol/handshake.rs [hunk 4]
```diff
@@ -42,55 +43,104 @@ impl SecWebSocket {
...
+                if let Some(ref location) = cfg.ar_location {
+                    req_header(format!("X-Algorand-Location: {location}"));
+                }
+                req_header(format!("X-Algorand-Noderandom: {}", cfg.ar_node_random));
+                if let Some(ref telid) = cfg.ar_tel_id {
+                    req_header(format!("X-Algorand-Telid: {telid}"));
+                }
+                req_header(format!("X-Algorand-Version: {}", cfg.ar_version));
+                req_header(format!("X-Algorand-Genesis: {}", cfg.ar_genesis));
+                req_header("".into()); // A HTTP header ends with '\r\n'
```
Comment (Rqnsom) [inline src/protocol/handshake.rs:138]: Every header must end with one `\r\n`, and then after the last header, we must append one more empty `\r\n`

> req_header doesn't add one at the end of the line.

req_header does add one at the end of the line? 🕵️‍♀️

File: src/protocol/handshake.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lwojtow) [reply src/protocol/handshake.rs:113]: Of course. Reminder to self: never do review before coffee. Still, confusing, non-obvious, etc...

[the remark this answers is not in the document]
Comment (Rqnsom) [reply src/protocol/handshake.rs:113]: > I would put \r\n second - after the string not before

`push_str` does exactly that 😅

> pub fn [push_str](https://doc.rust-lang.org/std/string/struct.String.html#method.push_str)(&mut self, [str](https://doc.rust-lang.org/std/primitive.str.html)ing: &str)
> 
> Appends a given string slice onto the end of this String.

**Review by asmie — APPROVED** [review]

File: src/tools/synthetic_node.rs [hunk 5]
```diff
@@ -50,7 +50,7 @@ impl Default for SyntheticNodeBuilder {
                 ..Default::default()
             },
             handshake: true,
-            challenge: None,
+            handshake_cfg: HandshakeCfg::default(),
```
[the remark this answers is not in the document]
Comment (Rqnsom) [reply src/tools/synthetic_node.rs]: yes! TIL!
Done

File: src/protocol/handshake.rs [hunk 6]
```diff
@@ -42,55 +43,104 @@ impl SecWebSocket {
...
+    /// Network priority challenge sent to clients which try to connect to the node.
+    pub challenge: Option<String>,
+    /// A key-accept pair for a Sec-WebSocket-Key header.
+    pub ws_key: Option<SecWebSocket>,
+}
+
+impl Default for HandshakeCfg {
+    fn default() -> Self {
+        Self {
+            ar_instance_name: String::from(X_AG_INSTANCE_NAME),
```
[the remark this answers is not in the document]
Comment (Rqnsom) [reply src/protocol/handshake.rs]: Done

**Review by Rqnsom — COMMENTED** [review]
> but need to clean clippy suggestions

my local Clippy is not showing these warnings 🕵️‍♀️

**Pull request merged**